src/ORM/DataObjectSchema.php

```diff
@@ -486,30 +494,72 @@ protected function cacheDatabaseFields($class)
     }
 
     /**
-     * Cache all indexes for the given class.
-     * Will do nothing if already cached
+     * Cache all indexes for the given class. Will do nothing if already cached.
      *
      * @param $class
      */
     protected function cacheDatabaseIndexes($class)
     {
-        if (array_key_exists($class, $this->databaseIndexes)) {
-            return;
+        if (!array_key_exists($class, $this->databaseIndexes)) {
+            $this->databaseIndexes[$class] = array_merge(
+                $this->cacheDefaultDatabaseIndexes($class),
+                $this->buildCustomDatabaseIndexes($class)
+            );
         }
+    }
+
+    /**
+     * Get "default" database indexable field types
+     *
+     * @param  string $class
+     * @return array
+     */
+    protected function cacheDefaultDatabaseIndexes($class)
+    {
         $indexes = [];
+        if (array_key_exists($class, $this->defaultDatabaseIndexes)) {
+            return $this->defaultDatabaseIndexes[$class];
+        }
 
-        // look for indexable field types
-        foreach ($this->databaseFields($class, false) as $field => $type) {
-            /** @skipUpgrade */
-            if ($type === 'ForeignKey' || $type === 'DBClassName') {
-                $indexes[$field] = [
+        $this->defaultDatabaseIndexes[$class] = [];
+
+        $fieldSpecs = $this->fieldSpecs($class, self::UNINHERITED);
+        foreach ($fieldSpecs as $field => $spec) {
+            list($fieldClass, $arguments) = ClassInfo::parse_class_spec($spec);
+            /** @var DBField $fieldObj */
+            $fieldObj = Injector::inst()->create($fieldClass, $field);
+
+            // Polymorphic indexes are handled as a DBComposite, check to avoid adding them twice
+            if ($fieldObj->config()->get('indexed') && array_shift($arguments) !== DataObject::class) {
```

Comment:
It's for catching polymorphics that have a class name added in - we don't need to index `PolymorphicClassName` on its own as well as the combination of PolymorphicID and *ClassName